drivers/sx126x/sx126x_radio_hal.c

```diff
@@ -0,0 +1,692 @@
+/*
+ * Copyright (C) ??? 2023 
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_sx126x
+ * @{
+ * @file
+ * @brief       Radio HAL implementation for the Semtech SX126x 
+ *
+ * @author      Klim Evdokimov <klimevdokimov@mail.ru>
+ * @}
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#include "net/ieee802154/radio.h"
+#include "event/callback.h"
+#include "event/thread.h"
+#include "ztimer.h"
+#include "checksum/ucrc16.h"
+#include "kernel_defines.h"
+#include "event.h"
+
+#include "sx126x.h"
+#include "sx126x_params.h"
+#include "sx126x_internal.h"
+
+#define LORA_ACK_REPLY_US          1024
+
+#ifndef SX126X_HAL_CHAN_BASE
+#define SX126X_HAL_CHAN_BASE (868300000LU)
+#endif
+
+#ifndef SX126X_HAL_CHAN_SPACING
+#define SX126X_HAL_CHAN_SPACING (200000LU)
+#endif
+
+#define SX126X_CHAN_MIN (IEEE802154_CHANNEL_MIN_SUBGHZ)
+
+#ifndef SX126X_CHAN_MAX
+#define SX126X_CHAN_MAX (8)
+#endif
+
+#define SX126X_POWER_MIN (-9)
+
+#ifndef SX126X_POWER_MAX
+#define SX126X_POWER_MAX (22)
+#endif
+
+
+static const ieee802154_radio_ops_t sx126x_ops;
+
+static int _set_state(sx126x_t *dev, sx126x_state_t state);
+static int _get_state(sx126x_t *dev, void* val);
+
+void _sx126x_handler(void* arg)
+{
+#if IS_USED(MODULE_IEEE802154)
+    ieee802154_dev_t *hal = arg;
+    sx126x_hal_task_handler(hal);
+#else
+    netdev_t *dev = arg;
+    netdev_trigger_event_isr(dev);
+    
+#endif
+}
+
+#if IS_USED(MODULE_SX126X_STM32WL)
+
+event_callback_t sx126x_ev_callback;
+#if IS_USED(MODULE_IEEE802154)
+void isr_subghz_radio(void)
+{
+    /* Disable NVIC to avoid ISR conflict in CPU. */
+    NVIC_DisableIRQ(SUBGHZ_Radio_IRQn);
+    NVIC_ClearPendingIRQ(SUBGHZ_Radio_IRQn);
+    event_post(EVENT_PRIO_MEDIUM, (event_t*)&sx126x_ev_callback);
+    cortexm_isr_end();
+}
+#endif
+#endif 
+
+void sx126x_hal_setup(sx126x_t *dev, ieee802154_dev_t *hal)
+{
+    hal->driver = &sx126x_ops;
+    hal->priv = dev;
+#if IS_USED(MODULE_SX126X_STM32WL)
+    event_callback_init(&sx126x_ev_callback, (void*)_sx126x_handler, (void*)hal);
+#endif
+}
+
+static void ack_timer_cb(void *arg)
```

Comment:
The transceiver is so slow, that you might consider offloading this using `event_post`.
This was not possible with e.g the `nrf802154`, as the reply should take less than 192 us.
Doing that will definitely prevent any synchronization issues